timelock/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+//! This contract demonstrates 'timelock' concept and implements a
+//! greatly simplified Claimable Balance (similar to
+//! https://developers.stellar.org/docs/glossary/claimable-balance).
+//! The contract allows to deposit some amount of token and allow another
+//! account(s) claim it before or after provided time point.
+#![no_std]
+#[cfg(feature = "testutils")]
+extern crate std;
+
+use soroban_auth::{
+    check_auth, NonceAuth, {Identifier, Signature},
+};
+use soroban_sdk::{contractimpl, contracttype, BigInt, BytesN, Env, IntoVal, Symbol, Vec};
+
+mod token {
+    soroban_sdk::contractimport!(file = "../soroban_token_contract.wasm");
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub enum DataKey {
+    Balance,
+    Nonce(Identifier),
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub enum TimeBoundKind {
+    Before,
+    After,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct TimeBound {
+    pub kind: TimeBoundKind,
+    pub timestamp: u64,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct ClaimableBalance {
+    pub token: BytesN<32>,
+    pub amount: BigInt,
+    pub claimants: Vec<Identifier>,
+    pub time_bound: TimeBound,
+}
+
+pub struct ClaimableBalanceContract;
+
+// The 'timelock' part: check that provided timestamp is before/after
+// the current ledger timestamp.
+fn check_time_bound(env: &Env, time_bound: &TimeBound) -> bool {
+    let ledger_timestamp = env.ledger().timestamp();
+
+    match time_bound.kind {
+        TimeBoundKind::Before => ledger_timestamp <= time_bound.timestamp,
+        TimeBoundKind::After => ledger_timestamp >= time_bound.timestamp,
+    }
+}
+
+// Contract usage pattern (pseudocode):
+// 1. Depositor calls `token.approve(depositor_auth, claimable_balance_contract, 100)`
+//    to allow contract to withdraw the needed amount of token.
+// 2. Depositor calls `deposit(depositor_auth, token_id, 100, claimants, time bound)`. Contract
+//    withdraws the provided token amount and stores it until one of the claimants
+//    claims it.
```

Comment:
It's unfortunate steps 1 and 2 have to be separate steps, but afaict it has to be this way without introducing significant complexity into the token contract and how it verifies signatures. cc @tamirms curious if you think there's other ways to do this that don't require two steps.

Reply:
@leighmcculloch The only way I can think of reducing the number of transactions is through something like [EIP 2612](https://eips.ethereum.org/EIPS/eip-2612) which introduces an extension to the erc20 standard that allows you to do the approval step off-chain through a signing process. There are [several tokens](https://help.1inch.io/en/articles/5435386-permit-712-signed-token-approvals-and-how-they-work-on-1inch) which have implemented  [EIP 2612](https://eips.ethereum.org/EIPS/eip-2612) including DAI and UNI.

Reply:
IIUC supporting call stack in auth would help resolving this - I feel like it's a useful feature to have in general.

Reply:
I don't think we need to do either of those options right now for this example, the example is good as is. But maybe this example is a good test bed for complex auth scenarios using the soroban-auth crate, once we resolves some of the issues we have open for it.